File: libs/mngr_llm/imbue/mngr_llm/resources/webchat_plugins/webchat_startup_notice.py
"""Startup notice endpoint for the webchat server.

Exposes ``GET /api/agent-startup-status`` which checks whether the
thinking agent's Claude Code session has started.  The companion
JavaScript plugin (``webchat_startup_notice.js``) polls this endpoint
and shows a banner when the session has not yet started, prompting
the user to connect via ``mngr connect`` to resolve any blocking
startup dialogs (e.g. the bypass-permissions confirmation).
"""

from __future__ import annotations

import os
from pathlib import Path
from typing import Final

from fastapi import FastAPI
from fastapi import Request
from fastapi.responses import JSONResponse
from llm_webchat.hookspecs import hookimpl
from pydantic import Field

from imbue.imbue_common.frozen_model import FrozenModel

_AGENT_STATE_DIR: Final[str] = os.environ.get("MNGR_AGENT_STATE_DIR", "")
_AGENT_NAME: Final[str] = os.environ.get("MNGR_AGENT_NAME", "")


def is_session_started(agent_state_dir: str) -> bool:
    """Check whether the Claude Code session has started.

    The SessionStart hook creates a ``session_started`` file in the agent
    state directory when the session begins.  Its absence means the agent
    is still blocked on a startup dialog.
    """
    if not agent_state_dir:
        return True
    return Path(agent_state_dir, "session_started").exists()


def _startup_status_endpoint(request: Request) -> JSONResponse:
    """Handler for GET /api/agent-startup-status."""
    agent_state_dir: str = request.app.state.startup_notice_state_dir
    agent_name: str = request.app.state.startup_notice_agent_name
    started = is_session_started(agent_state_dir)
    return JSONResponse(content={"started": started, "agent_name": agent_name})


class StartupNoticePlugin(FrozenModel):
    """Pluggy plugin that registers the /api/agent-startup-status endpoint."""

    agent_state_dir: str = Field(default=_AGENT_STATE_DIR, description="Agent state directory to check for session_started")
    agent_name: str = Field(default=_AGENT_NAME, description="Agent name shown in the connect command")

    @hookimpl
    def endpoint(self, app: FastAPI) -> None:
        app.state.startup_notice_state_dir = self.agent_state_dir
        app.state.startup_notice_agent_name = self.agent_name
        app.add_api_route(
            "/api/agent-startup-status",
            _startup_status_endpoint,
            methods=["GET"],
        )

File: libs/mngr_llm/imbue/mngr_llm/resources/webchat_plugins/webchat_startup_notice.js
/**
 * Startup notice plugin for llm-webchat.
 *
 * Polls GET /api/agent-startup-status to detect when the thinking agent
 * has not yet started its Claude Code session (e.g. because it is blocked
 * on a startup dialog like the bypass-permissions confirmation).  Shows a
 * dismissible banner at the top of the page with instructions to resolve
 * the issue via `mngr connect`.
 */
window.addEventListener("load", function () {
  "use strict";

  var POLL_INTERVAL_MS = 3000;
  var BANNER_ID = "startup-notice-banner";

  // ── Base path ────────────────────────────────────────────────

  function getBasePath() {
    var meta = document.querySelector('meta[name="llm-webchat-base-path"]');
    return ((meta && meta.getAttribute("content")) || "").replace(/\/+$/, "");
  }

  var basePath = getBasePath();

  // ── State ────────────────────────────────────────────────────

  var pollTimer = null;
  var dismissed = false;

  // ── DOM helpers ──────────────────────────────────────────────

  function escapeHtml(text) {
    var div = document.createElement("div");
    div.textContent = text;
    return div.innerHTML;
  }

  function removeBanner() {
    var existing = document.getElementById(BANNER_ID);
    if (existing) {
      existing.remove();
    }
  }

  function showBanner(agentName) {
    if (dismissed) return;
    if (document.getElementById(BANNER_ID)) return;

    var banner = document.createElement("div");
    banner.id = BANNER_ID;
    banner.style.cssText = [
      "position: fixed",
      "top: 0",
      "left: 0",
      "right: 0",
      "z-index: 10000",
      "background: #fff3cd",
      "color: #664d03",
      "padding: 10px 16px",
      "font-size: 14px",
      "font-family: system-ui, -apple-system, sans-serif",
      "display: flex",
      "align-items: center",
      "justify-content: space-between",
      "border-bottom: 1px solid #ffecb5",
      "box-shadow: 0 1px 3px rgba(0,0,0,0.1)",
    ].join("; ");

    var command = agentName ? "mngr connect " + escapeHtml(agentName) : "mngr connect <agent-name>";

    banner.innerHTML =
      '<span>The thinking agent is waiting at a startup prompt. Run ' +
      '<code style="background: rgba(0,0,0,0.08); padding: 2px 6px; border-radius: 3px; font-size: 13px;">' +
      command +
      "</code>" +
      " to resolve it.</span>" +
      '<button style="background: none; border: none; cursor: pointer; font-size: 18px; color: #664d03; padding: 0 4px; margin-left: 12px; line-height: 1;" title="Dismiss">&times;</button>';

    banner.querySelector("button").addEventListener("click", function () {
      dismissed = true;
      removeBanner();
    });

    document.body.appendChild(banner);
  }

  // ── Polling ──────────────────────────────────────────────────

  function poll() {
    fetch(basePath + "/api/agent-startup-status")
      .then(function (response) {
        if (!response.ok) return null;
        return response.json();
      })
      .then(function (data) {
        if (!data) return;

        if (data.started) {
          removeBanner();
          stopPolling();
        } else {
          showBanner(data.agent_name);
        }
      })
      .catch(function () {
        // Silently ignore network errors during polling.
      });
  }

  function startPolling() {
    poll();
    pollTimer = setInterval(poll, POLL_INTERVAL_MS);
  }

  function stopPolling() {
    if (pollTimer !== null) {
      clearInterval(pollTimer);
      pollTimer = null;
    }
  }

  // ── Initialization ───────────────────────────────────────────

  $llm.on("ready", function () {
    startPolling();
  });
});
